Exceptions.md

```diff
@@ -0,0 +1,199 @@
+
+The compiler will then compile all methods which return `Result<T>` into C# methods with signatures of the form
+
+```
+public T MethodCall(A1 a1, A2 a2, ...) throws Exception
+```
+
+(Note that C# method signatures do not have `throws` clauses, but we just add them here for clarity, even though that's not valid C# code).
+
+In particular, the same signature translation also applies for methods annotated with `{:extern}`, so for instance, given the C# method `System.IO.Stream.Read` with signature 
+
+```
+public int Read (byte[] buffer, int offset, int count)
+```
+
+one could write a Dafny stub for it as
+
+```
+method {:extern} Read(buffer: array<byte>, offset: int32, count: int32) returns (nRead: Result<int32>)
+```
+
+which would enable us to call exception-throwing C# methods from Dafny without an extra unverified layer turning exceptions into `Result<int32>`.
+
+In the same way, methods implemented in Dafny which return `Result<T>` would be compiled to exception-throwing methods and thus be directly usable by C# client programs.
+
+Moreover,
+
+```
+return Failure(someErrorMsg);
+```
+
+is compiled into a `throw` statement:
+
+```
+throw new Exception(dafnyStringToNativeString(someErrorMsg));
+```
+
+Compilation of
+
+```
+return Success(someValue);
+```
+
+simply drops the wrapping:
+
+```
+return someValue;
+```
+
+And the new Dafny construct
+
+```
+var value1: T1 :- operation1(args1);
+```
+
+is simply compiled into
+
+```
+var value1 = operation1(args1);
+```
+
+and the handling of the failure case is done by C#'s exception mechanism.
+
+
+## FAQ
+
+*Q:* Are we also adding special syntax/support for this inside expressions?
+*A:* Not at the moment
+
+*Q:* What about target languages which do not have exceptions, such as Go?
+*A:* Go uses a pattern of returning an additional return value to report success/failure, and the Dafny-to-Go compiler would use this pattern too. In general, the idea is to compile `Result<T>` to the idiomatic way of dealing with failure in the target language, be it exceptions, extra return values, `Option`, `Maybe`, ...
+
+*Q:* Can this work for methods which have multiple return values?
+*A:* It seems so.
+
+
+## Discussion points, open questions
+
+### Should  `Result<T>` be first-class?
+
+Should it be possible to pass around values of type `Result<T>`? That is, should it be allowed to assign a `Result<T>` to a variable, or to pass a `Result<T>` to a method? Say the user writes
+```
+var res1: Result<T1> := operation1(args1);
+anotherMethodCall(res1);
+```
+If this happens, and the compiler compiled `operation1` into a `public T1 operation1(args) throws Exception`, then the code generated for the above snippet would have to catch the exception and materialize it into a `Result<T1>`.
+In most cases, this is probably not the code which the user meant to write, so the questions are: Should this still be supported? Should it emit a warning? Should it emit an error?
+
```

Comment:
I like the idea of still supporting this, because it then seems that the Dafny code is more meaningful. That is, everything would work as you'd expect in Dafny, except if you start using the `:-` operator. But since this behavior would involve compiling to `try catch` blocks that do the switch from exceptions to a Dafny `Result<T>`, a warning may be good.

Also, something like:

    var res: Result<T> := ...;
    return res;

would compile into code like:

    var res: Result<T> := ...;  // no special handling here
    if res.Failure? {
      throw new Exception(dafnyStringToNativeString(res.error));
    } else {
      return res.value;
    }